263 update filter card css to make consistent spacing.

kartikeyakirar opened the pull request:
### closes https://github.com/insightsengineering/teal.slice/issues/263

Removed the margin and spaces between labels.

- Reduced the size of the checkbox label to make it same size as the group checkbox
- Removed margin for prettyswitch.

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/3f76d4a4-662c-4aaa-9298-aba70654ea92)

lcd2yyz commented:
While the margin/spacing looks good now, the font size looks different between the group checkbox labels vs keep NA. Could we fix that along the way? I would prefer to use the bigger font size as "Keep NA". 

In addition, there is an odd space before `Petal.Width`, not sure where that's coming from? I tried switching the order of the filters or adding other numerical variables, but only `Petal.Width` was affected.

![image](https://github.com/insightsengineering/teal.slice/assets/78242757/87987c95-b120-4074-9596-f0de0fa06d54)

m7pr commented:
@lcd2yyz great findings. Amazing we're having an effort to get the visual appearance as professional as possible

kartikeyakirar commented:
> In addition, there is an odd space before Petal.Width, not sure where that's coming from? I tried switching the order of the filters or adding other numerical variables, but only Petal.Width was affected.

The burst icon is prefixed to the title of petal.Width, causing the additional space. I'm not sure why it's not appearing for you. I checked by setting fixed to TRUE for another filter, and it had the same effect.
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/a98d2e8f-9a69-4f5e-802c-b6143bb988b8)

chlebowa commented:
Could it be another outdated dependency on BEE?

kartikeyakirar commented:
> While the margin/spacing looks good now, the font size looks different between the group checkbox labels vs keep NA. Could we fix that along the way? I would prefer to use the bigger font size as "Keep NA".

I have updated CSS and kept both font size to small . (`keep na` size is small)

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/609701dd-5b50-4401-bb02-dfd0e4907c06)

lcd2yyz commented:
> Also, looking through the css file, I notice that font sizes are specified with different units in different circumstances. I know this is warranted in some cases but perhaps a css overhaul would be beneficial. We talked about taking a holistic look at the `teal` UI/UX some time ago, can we add this to the agenda @lcd2yyz?

Okay I see. I thought it would be an easy fix. But if not, I'm fine with leaving it for now, and fix during the UX review altogether. Thanks for looking into it!

chlebowa commented:
> > Also, looking through the css file, I notice that font sizes are specified with different units in different circumstances. I know this is warranted in some cases but perhaps a css overhaul would be beneficial. We talked about taking a holistic look at the `teal` UI/UX some time ago, can we add this to the agenda @lcd2yyz?
> 
> Okay I see. I thought it would be an easy fix. But if not, I'm fine with leaving it for now, and fix during the UX review altogether. Thanks for looking into it!

Sorry, to be clear, I think the issue at hand **is** is fixed, and nice work, in fact.

kartikeyakirar commented:
@chlebowa 

> I now noticed that the variable label in the first card gets partially covered by the disable switch. This exact thing will not be a problem for long since branch 135 removes those switches altogether, but the underlying issue remains: the div containing the switch and the x mark (filter_cared_controls) is rendered on top of the div containing the header (filer_card_title).
> This should be fixed, but perhaps in a separate issue. You decide, @kartikeyakirar.
I believe we should address this as a separate issue because the overlapping occurs only when the screen is minimized to a specific size. To facilitate easy copy-paste, I captured screenshots while keeping the browser window at 50%. Here is the full-screen view.
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/b34b4e08-0202-4127-9c2f-407c89860761)
 and small screen view
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/843b9ff7-8160-41fa-8beb-4a8756fe8699)

We should consider discussing the expected behavior across all screen sizes.

chlebowa commented on inst/css/filter-panel.css:
Some margins are `0` and others are `0em`. Let's keep it consistent, pick one and stick to it.

kartikeyakirar replied:
I have replaced all the 0 margin with 0em and only refactored this.

kartikeyakirar replied:
Fixed the alignment issue. also fixed the spaces inconsistency between checkboxes
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/324e959b-e1fc-4fa2-920d-35a1ed8ec387)

chlebowa reviewed (changes requested):
![image](https://github.com/insightsengineering/teal.slice/assets/114988527/8c12e8ee-0a73-4575-8c36-70d12d1a9cc7)

It seems to me there are two cases of misalignment:
+ Keep NA text is aligned to bottom of checkbox, choice labels are aligned to center of checkboxes.
+ Keep NA text and choices labels are not left aligned.

kartikeyakirar commented:
> I now noticed that the variable label in the first card gets partially covered by the disable switch. This exact thing will not be a problem for long since branch 135 removes those switches altogether, but the underlying issue remains: the div containing the switch and the x mark (filter_cared_controls) is rendered on top of the div containing the header (filer_card_title).
> This should be fixed, but perhaps in a separate issue. You decide, @kartikeyakirar.
> 
> I believe we should address this as a separate issue because the overlapping occurs only when the screen is minimized to a specific size. To facilitate easy copy-paste, I captured screenshots while keeping the browser window at 50%. Here is the full-screen view.

I have added a separate issue i.e https://github.com/insightsengineering/teal.slice/issues/336

chlebowa reviewed (commented):
![image](https://github.com/insightsengineering/teal.slice/assets/114988527/04e90c87-4423-4a5c-965e-9a9829b30d60)

Now it seems that each text label is aligned to the top of the respective checkbox. Is it possible to have them aligned to the middle (center)?

not like this
![bad](https://github.com/insightsengineering/teal.slice/assets/114988527/ba9bf113-0a72-4f5f-94ae-be46ada3078d)
but like this
![good](https://github.com/insightsengineering/teal.slice/assets/114988527/e90b73fe-4f38-431d-bdc1-2f6bb86e36c7)

kartikeyakirar commented:
> Now it seems that each text label is aligned to the top of the respective checkbox. Is it possible to have them aligned to the middle (center)?

I am facing difficulty in vertically adjusting the position of the "keep NA" text within the widget. Despite checking the grouping and ensuring that the checkbox is placed in the middle relative to the progress state-bar-container div(), I am unable to achieve the desired alignment.

> ![image](https://github.com/insightsengineering/teal.slice/assets/6700955/c29566e9-6258-473b-b7d3-e159a07f41ec)

Furthermore, It seems , this alignment issue is visible on zooming in the browser. Specifically, the alignment of the text appears slightly off when the widget is not at the default 100% zoom level.

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/856a8f0e-cc83-4ac8-9e2a-a0813cc08d64)





https://github.com/insightsengineering/teal.slice/assets/6700955/8ff4843e-3561-4fe5-84db-3de220ae0618

chlebowa commented:
Well, I saw it at 100% but magnified to make it more apparent. If it cannot be done, I am willing to accept the current state, but do create an issue for that so we can have another go in the future.

kartikeyakirar commented:
Added issue for text alignment
https://github.com/insightsengineering/teal.slice/issues/338

chlebowa commented on inst/css/filter-panel.css:
```diff
@@ -178,6 +182,25 @@ a.remove_all:hover {
+  margin: 0em;
+}
+
+/* Needed for bs4 */
+.filterRangeSlider {
+  width: 100%;
+  max-width: 100%;
+}
+
```
```suggestion
```
No longer needed.

chlebowa reviewed (commented):
I think something went wrong after merging `filter_panel_refactor@main`, the separation between choices and keep NA is back:
![image](https://github.com/insightsengineering/teal.slice/assets/114988527/b59d8da7-80bb-4976-ae50-c9566e671bd8)

kartikeyakirar commented:
The alignment issue arises from the inclusion of a specific property within the `choices_state` class. This property is essential for the plotly range filter and cannot be removed from other div() as it impacts other properties. To address this, I have introduced an additional class called` range_choices_filter`, which includes a bottom margin to preserve the required layout.
here are the testing results on other filters
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/00896f22-99df-432a-8778-905f0afb7a4e)
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/e6c0d4a2-c2f6-4566-b97f-0a6647ea865f)
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/1c65bb0f-dfb1-4cf7-8eab-c50cbb3db1ad)
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/a5c738d2-1e92-4a3f-8bbe-8cb6931a02f3)

kartikeyakirar commented:
Additional note: I was not able to shift the text but I was able to shift the checkbox with new CSS selector for https://github.com/insightsengineering/teal.slice/issues/338 . @chlebowa Do you think its more aligned to the center now?

kartikeyakirar commented:
> I don't see a marked difference regarding the vertical alignment of checkbox and text.

yes, I can see this is not visible in the screenshot. I adjusted margin-top from  0.35em to 0.4em and it does look aligned in my browser.

kartikeyakirar commented:
@chlebowa @gogonzo  @lcd2yyz 

#### After incorporating ui_input within the` div` element with the class `choices_state`, and removing the `range_choices_state` while restoring the CSS property of choices_state, the changes added the margin after checkbox which in my opinion does not look bad at all. Here are screenshots displaying all the widgets.

- For plotly range

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/b1fefccc-4390-4b63-9e74-693890e502de)

- for plotly range with numeric input

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/d53e7805-d587-49b7-8eac-fb547c4f1130)

- Group checkbox with the keep-na checkbox

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/832d6788-bbc4-4c5f-b517-2e424dad98b8)

- radio button with the keep-na checkbox

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/0b09605a-aaca-47de-8eda-7688b1404b05)
-------------------------------------------------------------------------------
#### Another alternative is to revert the changes made on choice_state css class i.e retaining `margin-bottom: 0rem;` in choices_state. it will affect plotly range graph though. here is screenhots for comparision.


- For plotly range

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/62b20ba6-4c1f-46b3-8a1e-63157b58b264)

- for plotly range with numeric input

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/9e58ba33-ee02-48d3-a8eb-f2fce6c8c7c7)

- Group checkbox with the keep-na checkbox

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/35d6261c-8ef1-444b-ab5e-c90325f5c71a)

- radio button with the keep-na checkbox

![image](https://github.com/insightsengineering/teal.slice/assets/6700955/3b8cf1ce-ca2c-406f-82f5-b2286530db88)

kartikeyakirar commented:
Here are the discussed options, and for the first two options, I have included screenshots in the previous comment. Please let me know which option seems more appropriate:

1. Retaining the margin between the group checkbox/radio button and the "keep-NA" checkbox. 

2. Remove the bottom margin in the Plotly range, resulting in no spacing between the group checkbox/radio button and the "keep-NA" checkbox.

3. Stick with the current solution where an additional class has been added to override the choice_state property for rangeFilterState.

chlebowa commented:
I like the current state the best:
+ no `range_choices_state` class
+ `choices_state` has `margin-bottom: 0.45em`

The box-text alignment in Keep NA is still of on my machine (see screenshot) but there is already a separate issue for that.

Let's keep this!

![image](https://github.com/insightsengineering/teal.slice/assets/114988527/a807fa16-14f8-4502-84ad-72ecbeefce4c)

gogonzo reviewed (changes requested):
I guess that all the screenshots have been made with `options("teal.bs_theme" = bslib::bs_theme(version = 4))`. With generic shiny styling keepna and keeping looks bad


<img width="313" alt="Skärmavbild 2023-06-18 kl  23 46 10" src="https://github.com/insightsengineering/teal.slice/assets/6959016/bcf1a438-d57d-4018-90af-e042317f4e56">

```r
pkgload::load_all("teal")
pkgload::load_all("teal.slice")

funny_module <- function (label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <-  renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$numeric <- runif(nrow(ADSL))
ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- Inf

default_filters <- teal::teal_filters(
  filter_var(dataname = "ADSL", varname = "numeric")
)

app <- init(
  data = cdisc_data(
    cdisc_dataset("ADSL", ADSL)
  ),
  modules = modules(
    funny_module("filter module")
  ),
  filter = default_filters
)

runApp(app)
```

kartikeyakirar commented:
@gogonzo 
I have reverted some changes and updated a few properties to fix alignment issues with regularl version .

there are small issues which I fixed in regular but it affects the small margin in the bs4/5 version e.g alighment of KeepNA with check box. 

Here is demo in regular and bs4 version.

https://github.com/insightsengineering/teal.slice/assets/6700955/0f00a4e1-2dba-446d-96ea-f840a177d408


**Code used for testing**
```
options("teal.bs_theme" = bslib::bs_theme(version = 5))

funny_module <- function (label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <-  renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

options(teal.log_level = "TRACE", teal.show_js_log = F)
#options(shiny.trace = TRUE)
devtools::load_all()
library(teal)
library(scda)

Iris <- iris
Iris[["logical"]] <- sample(c(T,F), nrow(Iris),replace = T)
attr(Iris[[1]], "label") <- "variable label"
attr(Iris[[2]], "label") <- "variable label"
attr(Iris[[3]], "label") <- "variable label"
attr(Iris[[4]], "label") <- "variable label"

Iris$Species1 <- Iris$Species
Iris$Species1[sample(1:150,size = 40)] <- NA
data <- teal_data(
  dataset("iris", Iris),
  dataset("mtcars", mtcars)
)
Iris$Numeric_range <- 1:nrow(Iris)
Iris$Numeric_range[c(2,4)] <- Inf
Iris$Numeric_range[c(8:13)] <- NA
Iris[["logical1"]] <- sample(c(T,F), nrow(Iris),replace = T)
Iris$logical1[sample(1:150,size = 40)] <- NA
Iris$Species1[c(2,4)] <- Inf
data <- teal_data(
  dataset("iris", Iris),
  dataset("mtcars", mtcars)
)
default_filters <- filter_settings(
  filter_var("iris", "Petal.Width", sel = c(0.3, 1.8), fixed = TRUE, disabled = TRUE),
  filter_var("iris", "Numeric_range", sel = c(10, 20), fixed = F, disabled = FALSE),
  filter_var("iris", "Species", sel = c("versicolor", "virginica"), fixed = FALSE),
  filter_var("iris", "Species1", sel = c("versicolor", "virginica"), fixed = FALSE),
  filter_var("iris", "logical",  multiple=F),
  filter_var("iris", "logical1",  multiple=F),
  exclude_varnames = list(iris = c("Sepal.Length", "Sepal.Width"))
)

app <- init(
  data = data,
  modules = list(funny_module(), funny_module("module2")),
  filter = default_filters
)


runApp(app)
```

gogonzo reviewed (commented):
Almost there. Everything looks good except when numeric contains NA and Inf 

| this branch | feature branch|
|-----|-----|
|<img width="332" alt="Skärmavbild 2023-06-19 kl  14 10 35" src="https://github.com/insightsengineering/teal.slice/assets/6959016/c830b9f3-3440-431f-8b31-610db6f034a7">| <img width="327" alt="Skärmavbild 2023-06-19 kl  14 11 50" src="https://github.com/insightsengineering/teal.slice/assets/6959016/9709c250-4e76-49ac-8a79-6717762d50ee"> |

```r
options("teal.bs_theme" = bslib::bs_theme(version = 5))
pkgload::load_all("teal")
pkgload::load_all("teal.slice")
library(scda)

funny_module <- function(label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <- renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$numeric <- runif(nrow(ADSL))
ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$numeric[sample(1:nrow(ADSL), size = 10)] <- Inf
ADSL$categorical <- rep(c("a", "b", "c"), length.out = nrow(ADSL))
ADSL$categorical[sample(1:nrow(ADSL), size = 10)] <- NA
ADSL$categorical2 <- rep(c("a", "b", "c"), length.out = nrow(ADSL))
ADSL$categorical2[sample(1:nrow(ADSL), size = 10)] <- NA

default_filters <- teal::teal_filters(
  filter_var(dataname = "ADSL", varname = "numeric"),
  filter_var(dataname = "ADSL", varname = "categorical", multiple = FALSE),
  filter_var(dataname = "ADSL", varname = "categorical2", multiple = TRUE)
)

app <- init(
  data = cdisc_data(
    cdisc_dataset("ADSL", ADSL)
  ),
  modules = modules(
    funny_module("filter module")
  ),
  filter = default_filters
)

runApp(app)
```

kartikeyakirar commented:
I intentionally modified this property to align the checkbox more to the left on larger screen sizes, deviating from the original design in the feature branch. However, I have now reverted it back to its previous state.
![image](https://github.com/insightsengineering/teal.slice/assets/6700955/0b2975b7-165f-4d6e-8fec-5f61da8b7647)

gogonzo reviewed (approved):
👍
